### PR TITLE
Fix the bugs by using Dict for the selected_elements

### PR DESCRIPTION
**Pull request by dou-du**
In order to fix the bugs mentioned in issue #4 , I decide to use a Dict (traitlets) for the selected_elements.

```python
widget.selected_elements = {"H":0, "O":1}
```
However, there is a issue for the Dict. 

```python
#This will not work
widget.selected_elements["H"] = 2
```

### CODE REVIEW

**Comment by dou-du** [comment]
@giovannipizzi 
Please make a review for this pull request.

File: src/widget.ts [hunk 1]
```diff
@@ -81,25 +81,24 @@ class MCPTableView extends DOMWidgetView {
   // http://codebeerstartups.com/2012/12/how-to-improve-templates-in-backbone-js-learning-backbone-js/
   tableTemplate =  _.template( '<% for (let elementRow of elementTable)'+
   ' { print("<div class=\'periodic-table-row\'>"); for (let elementName of elementRow)'+
-  ' { if ( (elementName === "") || (elementName == "*" ) ) { %>' +
+  ' { if ( (elementName === "") || (elementName == "*" ) || (elementName == "#" ) ) { %>' +
```
Comment (CasperWA) [inline src/widget.ts:84]: I guess `*` is not needed anymore due to the change above from `*` -> `#`?
```suggestion
  ' { if ( (elementName == "") || (elementName == "#" ) ) { %>' +
```

File: src/widget.ts [hunk 2]
```diff
@@ -81,25 +81,24 @@ class MCPTableView extends DOMWidgetView {
...
   '(! disabledElements.includes(elementName)) ) { print(" elementOn"); } %>" '+
   'style="background-color: <% if (disabledElements.includes(elementName)) {print(disabledColor)}' +
-  'else if (selectedElements.includes(elementName)) { i = selectedElements.indexOf(elementName); print(selectedColors[selectedStates[i]]);} else{print(noselectColor)} %>" '+
-  'title="state: <% if (selectedElements.includes(elementName)) { i = selectedElements.indexOf(elementName); print(selectedStates[i]);} '+
-  'else if (disabledElements.includes(elementName)){print("disabled");} else {print("noselcted");} %>" ><% '+
-  'print(displayNamesReplacements[elementName] || elementName); %></span>' +
+  'else if (selectedElements.includes(elementName)) { i = selectedElements.indexOf(elementName); print(selectedColors[selectedStates[i]]);} else{print(unselectedColor)} %>" '+
+  // 'title="state: <% if (selectedElements.includes(elementName)) { i = selectedElements.indexOf(elementName); print(selectedStates[i]);} '+
+  // 'else if (disabledElements.includes(elementName)){print("disabled");} else {print("unselected");} %>" ><% '+
```
Comment (CasperWA) [inline src/widget.ts:93]: Why the commenting out?

File: src/widget.ts [hunk 3]
```diff
@@ -117,49 +116,66 @@ class MCPTableView extends DOMWidgetView {
...
 
+      else{
+        return;
+      }
+
       // Update the model (send back data to python)
-      this.model.set('selected_elements', newList);
-      this.model.set('selected_states', newStatesList);
+      // I have to make some changes, since there is some issue
+      // for Dict in Traitlets, which cannot triggle the update
```
Comment (CasperWA) [inline src/widget.ts]:
```suggestion
      // for Dict in Traitlets, which cannot trigger the update
```

Comment (CasperWA) [inline src/widget.ts]: I don't believe this is true. I think you just need to set a default value of `{}` for the traitlet or something, right?

File: src/widget.ts [hunk 4]
```diff
@@ -117,49 +116,66 @@ class MCPTableView extends DOMWidgetView {
...
+      this.model.set('selected_elements', {"Du": 0});
```
Comment (CasperWA) [inline src/widget.ts:173]: Why this? Shouldn't it be set to the actually selected elements?

File: src/widget.ts [hunk 5]
```diff
@@ -186,17 +213,24 @@ class MCPTableView extends DOMWidgetView {
...
     if (changed) {
       //             Make a copy before setting
-      while (newSelectedElements.length > newSelectedStates.length){
-        newSelectedStates.push(0);
-      }
-      this.model.set('selected_elements', newSelectedElements);
-      this.model.set('selected_states', newSelectedStates);
+      // while (newSelectedElements.length > newSelectedStates.length){
+      //   newSelectedStates.push(0);
+      // };
```
Comment (CasperWA) [inline src/widget.ts:225]: Why the commenting out?

File: widget_periodictable/periodic_table.py [hunk 6]
```diff
@@ -9,8 +9,9 @@
 """
 
 from ipywidgets import DOMWidget
-from traitlets import Unicode, Int, List, Dict
+from traitlets import Unicode, Int, List, Dict, observe, validate, TraitError, Dict
 from ._frontend import module_name, module_version
+from copy import deepcopy
 
 class PTableWidget(DOMWidget):
     """TODO: Add docstring here
```
Comment (CasperWA) [inline widget_periodictable/periodic_table.py]:
```suggestion
    """Periodic Table Widget
```

File: widget_periodictable/periodic_table.py [hunk 7]
```diff
@@ -21,29 +22,78 @@ class PTableWidget(DOMWidget):
...
+        self.selected_elements = selected_elements
 
         if len(selected_colors) < states:
-            self.selected_colors = selected_colors + ["#a6cee3", "#b2df8a", "#fdbf6f", "#6a3d9a", "#b15928", "#e31a1c", "#1f78b4", "#33a02c", "#ff7f00", "#cab2d6", "#ffff99"]
+            additional_colors = ["#a6cee3", "#b2df8a", "#fdbf6f", "#6a3d9a", "#b15928", "#e31a1c", "#1f78b4", "#33a02c", "#ff7f00", "#cab2d6", "#ffff99"]
+            self.selected_colors = selected_colors + additional_colors * (1 + (states - len(selected_colors)) // len(additional_colors))
+            self.selected_colors = self.selected_colors[:states]
+
+
+    def set_element_state(self, elementName, state):
```
Comment (CasperWA) [inline widget_periodictable/periodic_table.py:59]: I don't think this function is ever used. Hence the functionality of the widget is non-existing at the moment.

**Review by CasperWA — COMMENTED** [review]
Great with some updates on this widget! :)

Although I still have some issues using it to its fullest potential.

**Comment by CasperWA** [comment]
> @giovannipizzi
> Please make a review for this pull request.

@dou-du you can assign people for review on the right hand side of this page (the "Reviewers" section).

File: examples/introduction.ipynb [hunk 8]
```diff
@@ -1,8 +1,15 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# The widget periodic table"
+   ]
+  },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 10,
```
Comment (giovannipizzi) [inline examples/introduction.ipynb]: Are you committing on purpose a run notebook?

File: postBuild [hunk 9]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
-jupyter labextension install @jupyter-widgets/jupyterlab-manager 
+pip install -i https://test.pypi.org/simple/ widget-periodictable
```
Comment (giovannipizzi) [inline postBuild:2]: Do you want to suggest to install from test-pypi?

File: src/widget.ts (hunk 3 above)
Comment (giovannipizzi) [reply src/widget.ts]: I think the problem is that if it was already `{}`, no message is sent back.

**Review by giovannipizzi — COMMENTED** [review]
Minor comments.
I couldn't check the code in detail, though. Can I check it on binder already, I guess?

**Comment by giovannipizzi** [comment]
I checked on Binder, seems to be working fine!

I think it would be also useful to have a function `widget.unselect_element(elementName)`, but it's not critical for now, there are various ways to achieve so, e.g. disable and reenable the element (even if it's inelegant) or things like
```python
tmp = widget.selected_elements.copy()
tmp.pop(elementName)
widget.selected_elements = tmp
```

File: src/widget.ts (hunk 1 above)
Comment (dou-du) [reply src/widget.ts:84]: I used both "#" and "*" to separate lanthanides and actinides group.

File: src/widget.ts (hunk 2 above)
Comment (dou-du) [reply src/widget.ts:93]: Giovanni suggested to remove the tooltips from the periodic table.

File: src/widget.ts (hunk 3 above)
Comment (dou-du) [reply src/widget.ts]: No only for empty dict. For example newStateList was {"H":1, "O": 2} then

```javascript
newSateList = {"H": 3, "O": 1}
```
will not trigger the update in the python end PTable.selected_elements will show  {"H":1, "O": 2}

File: src/widget.ts (hunk 4 above)
Comment (dou-du) [reply src/widget.ts:173]: This is a dirty trick to force the update for the PTable.selected_elements as mentioned above.

File: src/widget.ts (hunk 5 above)
Comment (dou-du) [reply src/widget.ts:225]: This is for previous code, when using list for the selected_elements and selected_states. For the dictionary, it does not need anymore.

File: widget_periodictable/periodic_table.py (hunk 7 above)
Comment (dou-du) [reply widget_periodictable/periodic_table.py:59]: This function is import to set the element states.

For example:
```python
PTable.selected_elements["H"] = 2
```
will not work. 

On can use this function to set individual element's state.

File: examples/introduction.ipynb (hunk 8 above)
Comment (dou-du) [reply examples/introduction.ipynb]: I just commit to save all change. This is not needed to be committed. I will change it back.

File: postBuild (hunk 9 above)
Comment (dou-du) [reply postBuild:2]: I just use it to test the widget-periodictable package. I have released a new version at PyPi. I will change it to PyPi.

**Comment by dou-du** [comment]
Thanks @CasperWA and @giovannipizzi 

I will have a final checking about the code and approve the pull request.

**Pull request merged**